File: backend/src/domain/sanction.ts
export interface SanctionData {
    id: string;
    penalty: number;
    licenseRef?: string;
    licenseId?: string;
    reason: string;
    issuedAt: Date;
    inspectorRef?: string;
    inspectorId?: string;
    randomness?: string;
}

export class Sanction {
    readonly id: string;
    readonly penalty: number;
    readonly licenseRef: string;
    readonly reason: string;
    readonly issuedAt: Date;
    readonly inspectorRef: string;
    readonly randomness: string;

    constructor(data: SanctionData) {
        if (!data.id || data.id.trim() === "") {
            throw new Error("Sanction ID cannot be empty");
        }

        if (data.penalty <= 0) {
            throw new Error("Penalty must be greater than 0");
        }

        const ref = data.licenseRef ?? data.licenseId;
        if (!ref || ref.trim() === "") {
            throw new Error("License reference cannot be empty");
        }

        if (!data.reason || data.reason.trim() === "") {
            throw new Error("Reason cannot be empty");
        }

        const inspRef = data.inspectorRef ?? data.inspectorId;
        if (!inspRef || inspRef.trim() === "") {
            throw new Error("Inspector reference cannot be empty");
        }

        this.id = data.id.trim();
        this.penalty = data.penalty;
        this.licenseRef = ref.trim();
        this.reason = data.reason.trim();
        this.issuedAt = data.issuedAt;
        this.inspectorRef = inspRef.trim();
        this.randomness = data.randomness ?? "";
    }

    get sanctionId(): string {
        return this.id;
    }

    get licenseId(): string {
        return this.licenseRef;
    }

    get inspectorId(): string {
        return this.inspectorRef;
    }
}

File: chaincode/domain/sanction.go
package domain

import (
	"errors"
	"strings"
	"time"
)

type Sanction struct {
	ID                 string    `json:"id"`
	LicenseRef         string    `json:"licenseRef"`
	SanctionCommitment string    `json:"sanctionCommitment"`
	IssuedAt           time.Time `json:"issuedAt"`
	InspectorRef       string    `json:"inspectorRef"`
	Version            int       `json:"version"`
}

func NewSanction(
	id string,
	licenseRef string,
	sanctionCommitment string,
	issuedAt time.Time,
	inspectorRef string,
	version int,
) (*Sanction, error) {
	if strings.TrimSpace(id) == "" {
		return nil, errors.New("ID cannot be empty")
	}

	if strings.TrimSpace(licenseRef) == "" {
		return nil, errors.New("LicenseRef cannot be empty")
	}

	if strings.TrimSpace(sanctionCommitment) == "" {
		return nil, errors.New("SanctionCommitment cannot be empty")
	}

	if issuedAt.IsZero() {
		return nil, errors.New("IssuedAt cannot be zero")
	}

	if strings.TrimSpace(inspectorRef) == "" {
		return nil, errors.New("InspectorRef cannot be empty")
	}

	if version < 1 {
		return nil, errors.New("Version must be greater than or equal to 1")
	}

	return &Sanction{
		ID:                 strings.TrimSpace(id),
		LicenseRef:         strings.TrimSpace(licenseRef),
		SanctionCommitment: strings.TrimSpace(sanctionCommitment),
		IssuedAt:           issuedAt,
		InspectorRef:       strings.TrimSpace(inspectorRef),
		Version:            version,
	}, nil
}
